### PR TITLE
Feature - REST API to list remote searches

### PR DESCRIPTION
**Pull request by edalex-ian**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included

##### Description of change

Just another simple endpoint, this time to get a list of Remote Searches.

This is very similar to #2512 , and as a result includes a degree of refactoring to keep things DRY and pave the way for any future `BaseEntity` work we do. To some degree there is a whole framework on the Java side for this, however on this side we're attempting to keep it more concrete and explicit.

Also snuck in a minor refactor to sort the list of properties in `LegacyGuice`. (One day we probably want to sit down and consider if this approach is really what we want to be doing - one big bag for all the DI.)

[contributor license agreement]: https://www.apereo.org/node/676
[commit guidelines]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/RemoteSearchResource.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  @GET
+  @ApiOperation(
+    value = "List available Remote (Federated) Searches",
+    notes =
+      "This endpoint is used to retrieve available Remote Searches and is secured by SEARCH_FEDERATED_SEARCH",
+    response = classOf[BaseEntitySummary],
+    responseContainer = "List"
+  )
+  def getAll: Response =
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/RemoteSearchResource.scala:25]: This is the key work of this PR.

File: Source/Plugins/Core/com.equella.core/src/com/tle/core/powersearch/PowerSearchPrivileges.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.tle.core.powersearch;
+
+public final class PowerSearchPrivileges {
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/src/com/tle/core/powersearch/PowerSearchPrivileges.java:3]: Following a pattern I saw used for Remote Searches to centralise the string literals for privileges.

File: Source/Plugins/Core/com.equella.core/src/com/tle/core/powersearch/PowerSearchService.java [hunk 3]
```diff
@@ -40,4 +40,11 @@
   List<BaseEntityLabel> listAllForSchema(long schemaId);
 
   List<BaseEntityLabel> listAllForSchema(Schema schema);
+
+  /**
+   * List all available Power Searches (Advanced Searches) assuming 'SEARCH_POWER_SEARCH' ACL.
+   *
+   * @return available power searches
+   */
+  List<PowerSearch> enumerateSearchable();
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/src/com/tle/core/powersearch/PowerSearchService.java:49]: It is handy (and DB efficient) to be able to get straight at the Base Entities. I saw this was done and secured with Remote Searches, and it nice supports the refactor for `BaseEntitySummary`.

File: Source/Plugins/Core/com.equella.core/src/com/tle/legacy/LegacyGuice.java [hunk 4]
```diff
@@ -96,149 +97,151 @@
 
 public class LegacyGuice extends AbstractModule {
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/src/com/tle/legacy/LegacyGuice.java:98]: Here's the sort I mentioned. In there I have also added `FederatedSearchService`.

**Review by edalex-ian — COMMENTED** [review]
Just some notes for reviewers.

**Review by PenghaiZhang — APPROVED** [review]
Have asked Ian a few questions and all were addressed. So looks good to me!

**Comment by edalex-ian** [comment]
Thanks @PenghaiZhang . Merging! :+1:

**Pull request merged**